executor/mem_reader.go

```diff
@@ -0,0 +1,456 @@
+	outputOffset := make([]int, 0, len(us.columns))
+	for _, col := range idxReader.outputColumns {
+		outputOffset = append(outputOffset, col.Index)
+	}
+	return &memIndexReader{
+		baseExecutor:     us.baseExecutor,
+		index:            idxReader.index,
+		table:            idxReader.table.Meta(),
+		kvRanges:         kvRanges,
+		desc:             us.desc,
+		conditions:       us.conditions,
+		addedRows:        make([][]types.Datum, 0, len(us.dirty.addedRows)),
+		retFieldTypes:    us.retTypes(),
+		outputOffset:     outputOffset,
+		handleBytes:      make([]byte, 0, 16),
+		memIdxHandles:    make(map[int64]struct{}, len(us.dirty.addedRows)),
+		belowHandleIndex: us.belowHandleIndex,
+	}
+}
+
+func (m *memIndexReader) getMemRows() ([][]types.Datum, error) {
+	tps := make([]*types.FieldType, 0, len(m.index.Columns)+1)
+	cols := m.table.Columns
+	for _, col := range m.index.Columns {
+		tps = append(tps, &cols[col.Offset].FieldType)
+	}
+	if m.table.PKIsHandle {
+		for _, col := range m.table.Columns {
+			if mysql.HasPriKeyFlag(col.Flag) {
+				tps = append(tps, &col.FieldType)
+				break
+			}
+		}
+	} else {
+		// ExtraHandle Column tp.
+		tps = append(tps, types.NewFieldType(mysql.TypeLonglong))
+	}
+
+	mutableRow := chunk.MutRowFromTypes(m.retFieldTypes)
+	err := iterTxnMemBuffer(m.ctx, m.kvRanges, func(key, value []byte) error {
+		data, err := m.decodeIndexKeyValue(key, value, tps)
+		if err != nil {
+			return err
+		}
+		handle := data[m.belowHandleIndex].GetInt64()
+		m.memIdxHandles[handle] = struct{}{}
+
+		mutableRow.SetDatums(data...)
+		matched, _, err := expression.EvalBool(m.ctx, m.conditions, mutableRow.ToRow())
+		if err != nil || !matched {
+			return err
+		}
+		m.addedRows = append(m.addedRows, data)
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+	// TODO: After refine `IterReverse`, remove below logic and use `IterReverse` when do reverse scan.
+	if m.desc {
+		reverseDatumSlice(m.addedRows)
+	}
+	return m.addedRows, nil
+}
+
+func (m *memIndexReader) getMemRowsHandle() ([]int64, error) {
+	pkTp := types.NewFieldType(mysql.TypeLonglong)
+	if m.table.PKIsHandle {
+		for _, col := range m.table.Columns {
+			if mysql.HasPriKeyFlag(col.Flag) {
+				pkTp = &col.FieldType
+				break
+			}
+		}
+	}
+	handles := make([]int64, 0, cap(m.addedRows))
+	err := iterTxnMemBuffer(m.ctx, m.kvRanges, func(key, value []byte) error {
+		handle, err := m.decodeIndexHandle(key, value, pkTp)
+		if err != nil {
+			return err
+		}
+		handles = append(handles, handle)
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	for _, h := range handles {
+		m.memIdxHandles[h] = struct{}{}
+	}
+	if m.desc {
+		for i, j := 0, len(handles)-1; i < j; i, j = i+1, j-1 {
+			handles[i], handles[j] = handles[j], handles[i]
+		}
+	}
+	return handles, nil
+}
+
+func (m *memIndexReader) decodeIndexKeyValue(key, value []byte, tps []*types.FieldType) ([]types.Datum, error) {
+	// this is from indexScanExec decodeIndexKV method.
```

Comment:
why not reuse the logic in decodeIndexKV.

Reply:
done.